File: AQUA_python/runAQUA.py
import numpy as np
import matplotlib.pyplot as plt
from AQUA import aqua


# Neuron parameters
C = 50
vr = -60
vt = -40
k = 1.5
a = 0.03
b = 1
c = -40
d = 10
vpeak = 25

# Simulation parameters
T = 10000  # ms
h = 0.001  # ms
I = np.concatenate([np.zeros(int(0.01 * round(T / h))), 158 * np.ones(int(0.0074 * round(T / h))), 10 * np.ones(int(0.9826 * T / h))])

# Simulate different conditions (no autapse, delayed autapse, instantaneous autapse)
v1, _, _ , _= aqua(T, h, I, vpeak, C, vr, vt, k, a, b, c, d, 0.0, 0, 0)
v2, _, _, st2 = aqua(T, h, I, vpeak, C, vr, vt, k, a, b, c, d, 0.2, 100, 5)
v3, _, _, st3 = aqua(T, h, I, vpeak, C, vr, vt, k, a, b, c, d, 0.2, 100, 0)

# Plot the results
time = np.arange(round(T/h)) * h / 1000  # time in seconds

plt.figure(3, figsize=(10, 8))
plt.subplot(3, 2, 1); plt.plot(time, v1); plt.xlim([.008 * T / 1000, .03 * T / 1000]); plt.ylim([-60, 30]); plt.title('no autapse')
plt.subplot(3, 2, 3); plt.plot(time, v2); plt.xlim([.008 * T / 1000, .03 * T / 1000]); plt.ylim([-60, 30]); plt.title('delayed autapse')
plt.subplot(3, 2, 5); plt.plot(time, v3); plt.xlim([.008 * T / 1000, .03 * T / 1000]); plt.ylim([-60, 30]); plt.title('instantaneous autapse')
plt.subplot(3, 2, 2); plt.plot(time, v1); plt.xlim([.97 * T / 1000, T / 1000]); plt.ylim([-60, 30]); plt.title('no autapse')
plt.subplot(3, 2, 4); plt.plot(time, v2); plt.xlim([.97 * T / 1000, T / 1000]); plt.ylim([-60, 30]); plt.title('delayed autapse')
plt.subplot(3, 2, 6); plt.plot(time, v3); plt.xlim([.97 * T / 1000, T / 1000]); plt.ylim([-60, 30]); plt.title('instantaneous autapse')
plt.xlabel('time [sec]'); plt.ylabel('membrane potential [mV]')
plt.tight_layout()

# Plot ISI (Interspike Interval) analysis
plt.figure(4, figsize=(10, 4))
isi2 = np.diff(st2)
isi3 = np.diff(st3)
plt.subplot(1, 2, 1); plt.plot(isi2[-1000+1:-1], isi2[-1000+2:],'b.'); plt.title('ISI Delayed Autapse')
plt.subplot(1, 2, 2); plt.plot(isi3[-1000+1:-1], isi3[-1000+2:], 'r.'); plt.title('ISI Instantaneous Autapse')
plt.tight_layout()

plt.show()

File: AQUA_matlab/RunAQUA.m
%% Simulating quadratic I&F neuron with adaptation and autapse

clc
clear
close all

addpath(genpath('../HelperFunc'));

%%% bursting

% C=50; vr=-60; vt=-40; k=1.5; % parameters used for IB
% a=0.03; b=1; c=-40; d=150; % neocortical pyramidal neurons
% vpeak=25; % spike cutoff
% % vpeak=vt; % spike cutoff
% 
% T=50000; h=.001; % time span and step (ms)
% % I=[zeros(1,0.1*round(T/h)),500*ones(1,0.9*round(T/h))];% pulse of input DC current
% % I=[zeros(1,0.1*round(T/h)),500*ones(1,0.01*round(T/h)),500*ones(1,0.89*round(T/h))];% pulse of input DC current
% I=[zeros(1,0.01*round(T/h)),680*ones(1,0.99*round(T/h))];% pulse of input DC current
% 
% e=0.0; f = 0; tau = 0;
% [v1,~,~,~] = aqua(T,h,I,vpeak,C,vr,vt,k,a,b,c,d,e,f,tau);
% e=0.2; f = 100; tau = 5;
% [v2,~,~,st2]= aqua(T,h,I,vpeak,C,vr,vt,k,a,b,c,d,e,f,tau);
% e=0.2; f = 100; tau = 0;
% [v3,~,~,st3] = aqua(T,h,I,vpeak,C,vr,vt,k,a,b,c,d,e,f,tau);
% 
% figure(1)
% subplot(3,2,1); plot(h*(1:round(T/h))/1000, v1); xlim([.008*T/1000 .03*T/1000]); ylim([-60 -35]);
% subplot(3,2,3); plot(h*(1:round(T/h))/1000, v2); xlim([.008*T/1000 .03*T/1000]); ylim([-60 -35]);
% subplot(3,2,5); plot(h*(1:round(T/h))/1000, v3); xlim([.008*T/1000 .03*T/1000]); ylim([-60 -35]);
% subplot(3,2,2); plot(h*(1:round(T/h))/1000, v1); xlim([.97*T/1000 T/1000]); ylim([-60 -35]);
% subplot(3,2,4); plot(h*(1:round(T/h))/1000, v2); xlim([.97*T/1000 T/1000]); ylim([-60 -35]);
% subplot(3,2,6); plot(h*(1:round(T/h))/1000, v3); xlim([.97*T/1000 T/1000]); ylim([-60 -35]);
% 
% subplot(3,2,1); title('no autapse');
% subplot(3,2,3); title('delayed autapse');
% subplot(3,2,5); title('instantaneous autapse'); xlabel('time [sec]'); ylabel('membrane potential [mV]');
% subplot(3,2,2); title('no autapse');
% subplot(3,2,4); title('delayed autapse');
% subplot(3,2,6); title('instantaneous autapse');
% 
% figure(2);
% subplot(1,2,1); isi2 = diff(st2); plot(isi2(end-1000+1:end-1),isi2(end-1000+2:end),'.');
% subplot(1,2,2); isi3 = diff(st3); plot(isi3(end-1000+1:end-1),isi3(end-1000+2:end),'.');

% 
% e=0.0; f = 0; tau = 0;
% [v1,~,~,~] = aqua(T,h,I,vpeak,C,vr,vt,k,a,b,c,d,e,f,tau,2);
% e=0.2; f = 100; tau = 5;
% [v22,~,~,st22]= aqua(T,h,I,vpeak,C,vr,vt,k,a,b,c,d,e,f,tau,2);
% e=0.2; f = 100; tau = 0;
% [v3,~,~,st32] = aqua(T,h,I,vpeak,C,vr,vt,k,a,b,c,d,e,f,tau,2);
% 
% figure(2);
% subplot(1,2,1); hold on; isi22 = diff(st22); plot(isi22(end-1000+1:end-1),isi22(end-1000+2:end),'.');
% subplot(1,2,2); hold on; isi32 = diff(st32); plot(isi32(end-1000+1:end-1),isi32(end-1000+2:end),'.');
% 
% figure;plot(h*(9000000+(1:100000)-10001)/1000,v2(9000000+(1:100000)-10001));
% hold on;plot(h*(9000000+(1:100000)-10001)/1000,v22(9000000+(1:100000)-10001));
% xlabel('time [sec]'); ylabel('[mV]');
% title('membrane potential diverges exponentially due to perturbation');

%%% persistent

C=50; vr=-60; vt=-40; k=1.5; % parameters used for IB
a=0.03; b=1; c=-40; d=10; % neocortical pyramidal neurons
vpeak=25; % spike cutoff
% vpeak=vt; % spike cutoff

T=10000; h=.001; % time span and step (ms)
% I=[zeros(1,0.1*round(T/h)),500*ones(1,0.9*round(T/h))];% pulse of input DC current
I=[0*ones(1,0.01*round(T/h)),158*ones(1,0.0074*round(T/h)),10*ones(1,0.9826*round(T/h))];
% I=[zeros(1,0.01*round(T/h)),680*ones(1,0.99*round(T/h))];% pulse of input DC current

e=0.0; f = 0; tau = 0;
[v1,~,~,~] = aqua(T,h,I,vpeak,C,vr,vt,k,a,b,c,d,e,f,tau);
e=0.2; f = 100; tau = 5; % delayed autapse
[v2,~,~,st2]= aqua(T,h,I,vpeak,C,vr,vt,k,a,b,c,d,e,f,tau);
e=0.2; f = 100; tau = 0; % instantaneous autapse
[v3,~,~,st3] = aqua(T,h,I,vpeak,C,vr,vt,k,a,b,c,d,e,f,tau);

figure(3)
subplot(3,2,1); plot(h*(1:round(T/h))/1000, v1); xlim([.008*T/1000 .03*T/1000]); ylim([-60 30]);
subplot(3,2,3); plot(h*(1:round(T/h))/1000, v2); xlim([.008*T/1000 .03*T/1000]); ylim([-60 30]);
subplot(3,2,5); plot(h*(1:round(T/h))/1000, v3); xlim([.008*T/1000 .03*T/1000]); ylim([-60 30]);
subplot(3,2,2); plot(h*(1:round(T/h))/1000, v1); xlim([.97*T/1000 T/1000]); ylim([-60 30]);
subplot(3,2,4); plot(h*(1:round(T/h))/1000, v2); xlim([.97*T/1000 T/1000]); ylim([-60 30]);
subplot(3,2,6); plot(h*(1:round(T/h))/1000, v3); xlim([.97*T/1000 T/1000]); ylim([-60 30]);

subplot(3,2,1); title('no autapse');
subplot(3,2,3); title('delayed autapse');
subplot(3,2,5); title('instantaneous autapse'); xlabel('time [sec]'); ylabel('membrane potential [mV]');
subplot(3,2,2); title('no autapse');
subplot(3,2,4); title('delayed autapse');
subplot(3,2,6); title('instantaneous autapse');

figure(4);
subplot(1,2,1); isi2 = diff(st2); plot(isi2(end-1000+1:end-1),isi2(end-1000+2:end),'.');
subplot(1,2,2); isi3 = diff(st3); plot(isi3(end-1000+1:end-1),isi3(end-1000+2:end),'.');

% [x,y,z] = lorenz(T,h,1,1,1,10,28,8/3);
% figure;plot3(x,y,z);

